packages/code-du-travail-data/indexing/addGlossary.js

```diff
@@ -0,0 +1,56 @@
+import { v4 as uuidv4 } from "uuid";
+const glossary = require("@socialgouv/datafiller-data/data/glossary.json");
+
+const conventionMatchers = [
+  "convention collective",
+  "conventions collectives",
+  "accords de branches",
+  "accord de branche",
+  "disposition conventionnelle",
+  "dispositions conventionnelles",
+];
+
+export function addGlossary(htmlContent) {
+  const internalRefMap = new Map();
+  const formatedContent = glossary.reduce((updatedContent, match) => {
+    const { abbrs, title, definition, variants } = match;
+    if (!updatedContent) return updatedContent;
+    // we cannot use \b word boundary since \w does not match diacritics
+    // So we do a kind of \b equivalent.
+    // the main différence is that matched pattern can include a whitespace as first char
+    const frDiacritics = "àâäçéèêëïîôöùûüÿœæÀÂÄÇÉÈÊËÎÏÔÖÙÛÜŸŒÆ";
+    const wordBoundaryStart = `(?:^|[^_/\\w${frDiacritics}-])`;
+    const wordBoundaryEnd = `(?![\\w${frDiacritics}])`;
+    const patterns = [...new Set([title, ...variants])].map(
+      (term) =>
+        new RegExp(`${wordBoundaryStart}(${term})${wordBoundaryEnd}`, "gi")
+    );
+    if (abbrs) {
+      patterns.concat(new RegExp(`\\b(${abbrs})\\b`, "g"));
+    }
+
+    const formatedContent = Array.from(patterns).reduce((content, pattern) => {
+      return content.replace(pattern, function (_, term) {
+        // we use an internal ref id to track pattern replacement
+        const ref = "__tt__" + uuidv4();
+        const webComponent = conventionMatchers.includes(term)
+          ? `<webcomponent-tooltip-cc>${term}</webcomponent-tooltip-cc>`
+          : `<webcomponent-tooltip content="${encodeURI(
+              definition
+            )}">${term}</webcomponent-tooltip>`;
+
+        internalRefMap.set(ref, webComponent);
+        return _.replace(new RegExp(term), ref);
+      });
+    }, updatedContent);
+    return formatedContent;
+  }, htmlContent);
+
+  // we get ref id to replace pattern
+  const finalContent = Array.from(internalRefMap).reduce((content, match) => {
+    const [key, value] = match;
+    return content.replace(new RegExp(key, "g"), value);
+  }, formatedContent);
+
+  return finalContent;
+}
```

Comment:
Est-ce que ce n'est pas un peu plus lisible sans les reduces imbriqués ?
```suggestion
export function addGlossary(htmlContent) {
  if (!htmlContent) return "";

  let formatedContent = htmlContent;
  const uuidToWebComponent = new Map();

  glossary.forEach(({ abbrs, title, definition, variants }) => {
    const patterns = [title, ...variants].map(
      (term) =>
        new RegExp(`${wordBoundaryStart}(${term})${wordBoundaryEnd}`, "gi")
    );
    if (abbrs) {
      patterns.concat(new RegExp(`\\b(${abbrs})\\b`, "g"));
    }

    patterns.forEach((pattern) => {
      // while we loop, we replace the matches with uuid to prevent nested matches
      formatedContent = formatedContent.replace(pattern, function (
        match, // contains the matching term with the word boundaries
        term // contains only the matching term
      ) {
        const uuid = "__tt__" + uuidv4();
        const webComponent = conventionMatchers.includes(term)
          ? `<webcomponent-tooltip-cc>${term}</webcomponent-tooltip-cc>`
          : `<webcomponent-tooltip content="${encodeURI(
              definition
            )}">${term}</webcomponent-tooltip>`;

        uuidToWebComponent.set(uuid, webComponent);
        return match.replace(new RegExp(term), uuid);
      });
    });
  });

  // In the end, we replace the uuid with its related component
  let finalContent = formatedContent;
  Array.from(uuidToWebComponent).forEach(([uuid, webComponent]) => {
    finalContent = finalContent.replace(new RegExp(uuid, "g"), webComponent);
  });

  return finalContent;
}
```

Reply:
dans le diff c'est illisible :D fait un gist plutot

Reply:
merde, je suis en split diff pour moi c'est lisible atta je change de view historie de me rendre compte

Reply:
Essaye avec ça in the meanwhile ?
<img width="536" alt="Screenshot 2020-04-30 at 11 44 04" src="https://user-images.githubusercontent.com/705453/80696636-f4547980-8ad7-11ea-841b-5f3429d1e070.png">

Reply:
Ah non ça change rien ! Je comprends pas :o

Reply:
https://gist.github.com/UnbearableBear/e12077e173f4e77b284fd903b984d06e

Reply:
Après si vraiment on veut être "lean" on peut sortir le `foreach` des patterns du `foreach` du glossary !

Reply:
fait un comment dans le gist, je vois pas bien

Reply:
done !

Reply:
J'ai rajouté un commentaire dans le gist sur l'ordre des elements dans le tableau pattern pour s'assurer que les gros termes soient bien pris en compte